Mathlib/Combinatorics/SimpleGraph/Connectivity.lean

```diff
@@ -1208,6 +1230,31 @@ theorem exists_boundary_dart {u v : V} (p : G.Walk u v) (S : Set V) (uS : u ∈
       exact ⟨d, List.Mem.tail _ hd, hcd⟩
     · exact ⟨⟨(x, y), a⟩, List.Mem.head _, uS, h⟩
 
+/-- Given a walk `w` and a node in the support, there exists a natural `n`, such that given node
+is the `n`-th node (zero-indexed) in the walk. In addition, `n` is at most the length of the path.
+Due to the definition of `getVert` it would otherwise be legal to return a larger `n` for the last
+node. -/
+theorem exists_getVert {u v w : V} (w : G.Walk v w) (h : u ∈ w.support) :
+    ∃ n, w.getVert n = u ∧ n ≤ w.length := by
+  obtain ⟨q, r, hqr⟩ := SimpleGraph.Walk.mem_support_iff_exists_append.mp h
+  use q.length
+  rw [hqr]
+  rw [Walk.getVert_append]
+  simp only [lt_self_iff_false, ↓reduceIte, Nat.sub_self, getVert_zero, length_append,
+    Nat.le_add_right, and_self]
+
+lemma getVert_tail_support_get {u v n} (p : G.Walk u v) (hnp: ¬ p.Nil) :
+    (p.tail hnp).getVert n = p.getVert (n + 1) :=
+  p.notNilRec (fun _ _ ↦ by simp only [cons_tail, cons_getVert_succ]) hnp
+
+@[simp]
+lemma cons_sndOfNotNil (q : G.Walk v w) (hadj : G.Adj u v) :
+    (Walk.cons hadj q).sndOfNotNil (not_nil_cons) = v := by
+  unfold sndOfNotNil; simp only [notNilRec_cons]
+
+lemma getVert_one (p : G.Walk u v) (hnp : ¬ p.Nil) : p.getVert 1 = p.sndOfNotNil hnp :=
```

Comment:
Given this I'm confused why `sndOfNotNil` exists at all. It seems one should always just use `getVert 1`, and then the non-nil hypothesis is only needed in theorems.

Reply:
This question is a little bit above my paygrade. My educated guess is that it makes it easier to write stuff like this: `lemma cons_tail_eq (p : G.Walk x y) (hp : ¬ p.Nil) : cons (p.adj_sndOfNotNil hp) (p.tail hp) = p`

@YaelDillies Do you have any insight?

Reply:
I actually agree with Kim here. The existence of `sndOfNotNil` is dubious.

You are very welcome to remove it, but another PR since it is out of scope for this one.

Reply:
Since you both agree it makes sense, I think this time would be a good time to do it, since I also have a lot of stuff lined up where I used `p.reverse.sndOfNotNil` and standardizing on `p.getVert (p.length - 1)` for that will probably simplify stuff a lot, so I'll put this on the backlog for me.